datanode/dehistory/snapshot/block_commit_handler.go

```diff
@@ -0,0 +1,83 @@
+package snapshot
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"time"
+
+	"code.vegaprotocol.io/vega/core/netparams"
+	"code.vegaprotocol.io/vega/datanode/broker"
+	"code.vegaprotocol.io/vega/datanode/entities"
+	"code.vegaprotocol.io/vega/datanode/sqlstore"
+	"code.vegaprotocol.io/vega/logging"
+)
+
+type BlockCommitHandler struct {
+	log                 *logging.Logger
+	snapshotData        func(ctx context.Context, chainID string, toHeight int64, fromHeight int64) error
+	getNetworkParameter func(ctx context.Context, key string) (entities.NetworkParameter, error)
+	blockInterval       int64
+	brokerConfig        broker.Config
+}
+
+func NewBlockCommitHandler(
+	log *logging.Logger,
+	snapshotData func(ctx context.Context, chainID string, toHeight int64, fromHeight int64) error,
+	getNetworkParameter func(ctx context.Context, key string) (entities.NetworkParameter, error),
+	brokerConfig broker.Config,
+) *BlockCommitHandler {
+	return &BlockCommitHandler{
+		log:                 log.Named("block-commit-handler"),
+		snapshotData:        snapshotData,
+		getNetworkParameter: getNetworkParameter,
+		brokerConfig:        brokerConfig,
+	}
+}
+
+func (b *BlockCommitHandler) OnBlockCommitted(ctx context.Context, chainID string, blockHeight int64) bool {
+	// We poll for the snapshot interval on block commit to ensure that the correct interval
+	// is always used, this is simpler than listening for the network parameter event and then
+	// having to handle recovery scenarios.
+	param, err := b.getNetworkParameter(ctx, netparams.SnapshotIntervalLength)
+	if err != nil {
+		if !errors.Is(err, sqlstore.ErrNoParameterFound) {
+			b.log.Errorf("failed to get snapshot interval length network parameter:%w", err)
+		}
+	} else {
+		blockInterval, err := strconv.ParseInt(param.Value, 10, 64)
+		if err != nil {
+			b.log.Errorf("failed to parse snapshot interval length network parameter:%w", err)
+		} else {
+			b.blockInterval = blockInterval
+
+			// An interval less than 1000 when using a file source with no time between blocks results
+			// in excessive snapshot data creation and should be avoided, 1000 is a reasonable default
+			if b.brokerConfig.UseEventFile &&
+				b.brokerConfig.FileEventSourceConfig.TimeBetweenBlocks.Duration < time.Second {
+				if blockInterval < 1000 {
+					b.blockInterval = 1000
+				}
+			}
+		}
+	}
+
+	if b.snapshotRequiredAtBlockHeight(blockHeight) {
+		fromHeight := GetFromHeight(blockHeight, b.blockInterval)
+		b.snapshotData(ctx, chainID, blockHeight, fromHeight)
+	}
+
+	return false
```

Comment:
This seems to be the only way out of this function, perhaps we should remove the return arg?

Reply:
yeah, sorted this and remove that bool return val, seems it was only being used in tests.  now uses the context instead to cause the broker to exit